From 440f52e95613ee9fe82bfe1cffe7b9390c8c33c6 Mon Sep 17 00:00:00 2001
From: Jan Sondhauss <jan.sondhauss@wago.com>
Date: Thu, 14 Jun 2018 11:10:12 +0200
Subject: [PATCH] dsa: marvell: add defaul vlan for profinet ports

we don't want the profinet frames to egress from all switch ports
when learning. so in addition to the port based vlan, we add
a vlan where only profinet ports are participating. (port 3,4,6)

this is needed because we bypass the dsa/bridge layer for profinet
frames.
---
 drivers/net/dsa/mv88e6xxx/chip.c | 29 ++++++++++++++++++++++++++++-
 1 file changed, 28 insertions(+), 1 deletion(-)

diff --git a/drivers/net/dsa/mv88e6xxx/chip.c b/drivers/net/dsa/mv88e6xxx/chip.c
index 37cc302..1f1d996 100644
--- a/drivers/net/dsa/mv88e6xxx/chip.c
+++ b/drivers/net/dsa/mv88e6xxx/chip.c
@@ -2456,6 +2456,23 @@ static int mv88e6xxx_serdes_power_on(struct mv88e6xxx_chip *chip)
 	return err;
 }
 
+static int mv88e6xxx_vtu_setup(struct mv88e6xxx_chip *chip)
+{
+	int err = 0;
+
+	struct mv88e6xxx_vtu_entry pnio_vlan = {
+		.vid = 1337,
+		.fid = 0,
+		.sid = 0,
+		.valid = true,
+		.data = {3, 3, 3, 0, 0, 3, 0}
+	};
+
+	err = _mv88e6xxx_vtu_loadpurge(chip, &pnio_vlan);
+
+	return err;
+}
+
 static int mv88e6xxx_setup_port(struct mv88e6xxx_chip *chip, int port)
 {
 	struct dsa_switch *ds = chip->ds;
@@ -2591,7 +2608,10 @@ static int mv88e6xxx_setup_port(struct mv88e6xxx_chip *chip, int port)
 			reg |= PORT_CONTROL_2_FORWARD_UNKNOWN;
 	}
 
-	reg |= PORT_CONTROL_2_8021Q_DISABLED;
+	if (dsa_is_cpu_port(ds, port))
+		reg |= PORT_CONTROL_2_8021Q_FALLBACK;
+	else
+		reg |= PORT_CONTROL_2_8021Q_DISABLED;
 
 	if (reg) {
 		err = mv88e6xxx_port_write(chip, port, PORT_CONTROL_2, reg);
@@ -2907,6 +2927,13 @@ static int mv88e6xxx_setup(struct dsa_switch *ds)
 			goto unlock;
 	}
 
+	if (mv88e6xxx_6320_family(chip)) {
+		err = mv88e6xxx_vtu_setup(chip);
+		if (err)
+			goto unlock;
+	}
+
+
 unlock:
 	mutex_unlock(&chip->reg_lock);
 
-- 
2.7.4

